pkg/controllers/provisioningv2/fleetworkspace/controller_test.go

```diff
@@ -0,0 +1,188 @@
+package fleetworkspace
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/golang/mock/gomock"
+	v3 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/controllers/managementagent/nslabels"
+	mgmtcontrollers "github.com/rancher/rancher/pkg/generated/controllers/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/project"
+	"github.com/rancher/wrangler/pkg/generic/fake"
+	"github.com/stretchr/testify/assert"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+func TestOnChange(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	fleetWorkspacesProject := v3.Project{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "FleetWorkspaces",
+		},
+		Spec: v3.ProjectSpec{
+			ClusterName: "local",
+		},
+	}
+	workspace := v3.FleetWorkspace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "workspace",
+		},
+	}
+
+	tests := map[string]struct {
+		projectsMock       func() mgmtcontrollers.ProjectCache
+		expectedNamespaces []runtime.Object
+		expectedErr        error
+	}{
+		"FleetWorkspaces project exists": {
+			projectsMock: func() mgmtcontrollers.ProjectCache {
+				projectsMock := fake.NewMockCacheInterface[*v3.Project](ctrl)
+				projectsMock.EXPECT().List("local", labels.Set(project.FleetWorkspacesProjectLabel).AsSelector()).Return([]*v3.Project{&fleetWorkspacesProject}, nil)
+
+				return projectsMock
+			},
+			expectedNamespaces: []runtime.Object{
+				&corev1.Namespace{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: workspace.Name,
+						Annotations: map[string]string{
+							nslabels.ProjectIDFieldLabel: fmt.Sprintf("%v:%v", fleetWorkspacesProject.Spec.ClusterName, fleetWorkspacesProject.Name),
+						},
+						Labels: map[string]string{},
+					},
+				},
+			},
+		},
+		"FleetWorkspaces project does not exists": {
```

Comment:
Could you add a test where List returns an error as well?

Reply:
added in https://github.com/rancher/rancher/pull/44262/commits/2e9a9057eb5ebb838a1bd4399a37c6dd08057d08